lua-mode.el

```diff
@@ -1718,17 +1765,64 @@ When called interactively, switch to the process buffer."
   (interactive)
   (set-marker lua-region-end (or arg (point))))
 
-(defun lua-send-string (str)
-  "Send STR plus a newline to the Lua process.
-
-If `lua-process' is nil or dead, start a new process first."
-  (unless (string-equal (substring str -1) "\n")
-    (setq str (concat str "\n")))
-  (process-send-string (lua-get-create-process) str))
+(defvar lua-shell-temp-file nil
+  "Absolute pathname for temporary lua file for dofile'ing regions/long commands")
+
+(defun lua-shell-temp-file ()
+  "Returns a temp file for running lua commands, creating it if necessary"
+  (or lua-shell-temp-file
+      (setq lua-shell-temp-file
+	      (make-temp-file "lua-command-"))))
+
+(defun lua-shell-delete-temp-file ()
+  "Delete the temporary file."
+  (if (stringp lua-shell-temp-file)
+      (condition-case nil
+	  (delete-file lua-shell-temp-file)
+	(error nil))))
+(add-hook 'kill-buffer-hook 'lua-shell-delete-temp-file nil 'local)
+(add-hook 'kill-emacs-hook 'lua-shell-delete-temp-file)
+
+(defun lua-send-command-output-to-buffer-and-wait (command)
+  "Send a command accumulating output in the output buffer,
+and wait for the next prompt to appear. Blocks emacs, only use
+for instances where results are immediately needed.  Any lua
+output will be left in lua-shell-redirected-output."
+  (let ((process (lua-get-create-process)))
+    (with-current-buffer lua-process-buffer
+      (unless (buffer-live-p lua-shell-output-buffer)
+	(setq lua-shell-output-buffer
+	      (get-buffer-create
+	       (generate-new-buffer lua-shell-output-buffer-name))))
+      (with-current-buffer lua-shell-output-buffer (erase-buffer))
+      (lua-send-string command lua-shell-output-buffer process)
+      (while (not comint-redirect-completed)
+	(accept-process-output process))
+      (comint-redirect-cleanup))))
+
+(defun lua-send-string (str &optional redirect-buffer process)
+  "Send STR to the Lua process, possibly via dofile.
+If necessary, start a new process first.  If optional argument
+REDIRECT-BUFFER is passed, redirect command output to that
+buffer, which calls comint-redirect-hook when the output is
+complete.  If PROCESS not passed, get or create a process."
+  (let (file
+	(command str)
+	(process (or process (lua-get-create-process)))
+	(comint-redirect-perform-sanity-check nil))
+    (when (> (length str) lua-shell-maximum-command-length)
```

Comment:
That's right.  My testing indicated that there is a maximum command length, and in fact the length of the temp file path on a given emacs install was one explanation for why tests would confusingly fail on Travis while working on our local systems. But I never found this documented.  If indeed there is a maximum length, this could be useful for future interactions with the shell of any kind.

Reply:
What I'm saying that `lua-shell-maximum-command-length` as variable name could be interpreted as the limit to what the user can enter in Lua shell, whereas it is a limit specific to interprocess communication, to which there is already a solution. So now it's really a threshold after which the command is no longer sent via stdin but via temp-file, hence I'm wondering if there's a better name that makes sense to you.

Reply:
I switched to `lua-shell-maximum-sent-length` and documented as specific to Emacs sending commands to Lua (which I'm not sure if true but better represents our state of knowledge).

Reply:
:+1: